ddtrace/profiling/exporter/_packages.py

```diff
@@ -0,0 +1,126 @@
+# -*- encoding: utf-8 -*-
+import enum
+import logging
+import os
+import sys
+
+
+try:
+    import pathlib
+except ImportError:
+    import pathlib2 as pathlib  # type: ignore[no-redef]
+
+try:
+    import importlib.metadata as il_md
+except ImportError:
+    import importlib_metadata as il_md  # type: ignore[no-redef]
+
+import typing
+
+
+LOG = logging.getLogger(__name__)
+
+
+try:
+    fspath = os.fspath
+except AttributeError:
+
+    # Stolen from Python 3.10
+    def fspath(path):
+        # For testing purposes, make sure the function is available when the C
+        # implementation exists.
+        """Return the path representation of a path-like object.
+
+        If str or bytes is passed in, it is returned unchanged. Otherwise the
+        os.PathLike interface is used to get the path representation. If the
+        path representation is not str or bytes, TypeError is raised. If the
+        provided path is not str, bytes, or os.PathLike, TypeError is raised.
+        """
+        if isinstance(path, (str, bytes)):
+            return path
+
+        # Hack for Python 3.5: there's no __fspath__ :(
+        if sys.version_info[:2] == (3, 5) and isinstance(path, pathlib.Path):
+            return str(path)
+
+        # Work from the object's type to match method resolution of other magic
+        # methods.
+        path_type = type(path)
+        try:
+            path_repr = path_type.__fspath__(path)
+        except AttributeError:
+            if hasattr(path_type, "__fspath__"):
+                raise
+            else:
+                raise TypeError("expected str, bytes or os.PathLike object, not " + path_type.__name__)
+        if isinstance(path_repr, (str, bytes)):
+            return path_repr
+        else:
+            raise TypeError(
+                "expected {}.__fspath__() to return str or bytes, "
+                "not {}".format(path_type.__name__, type(path_repr).__name__)
+            )
+
+
+# We don't store every file of every package but filter commonly used extensions
+SUPPORTED_EXTENSIONS = (".py", ".so", ".dll", ".pyc")
+
+
+Distribution = typing.NamedTuple("Distribution", [("name", str), ("version", str)])
+
+
+def _is_python_source_file(
+    path,  # type: pathlib.PurePath
+):
+    # type: (...) -> bool
+    for ext in SUPPORTED_EXTENSIONS:
+        if path.name.endswith(ext):
+            return True
```

Comment:
would it be "better" (more performant?) to do `os.path.splitext(path.name) in SUPPORTED_EXTENSIONS`?

I suppose it doesn't matter too much since this is a one time on startup process, but if there are a _ton_ of files, maybe it'll matter?

Reply:
Yes, that's an excellent suggestion.